sql/core/src/test/scala/org/apache/spark/sql/DataFrameWindowFunctionsSuite.scala

```diff
@@ -1276,6 +1276,147 @@ class DataFrameWindowFunctionsSuite extends QueryTest
+              Seq(
+                Row("a", 0, "c", 4),
+                Row("a", 1, "x", 3),
+                Row("a", 2, "y", 2),
+                Row("a", 3, "z", 1),
+                Row("a", 4, "", 5),
+                Row("a", 4, "", 5),
+                Row("b", 1, "n", 1)
+              )
+            )
+            checkAnswer(df.withColumn("rn", dense_rank().over(window3)).limit(11),
+              Seq(
+                Row("a", 0, "c", 4),
+                Row("a", 1, "x", 7),
+                Row("a", 2, "y", 8),
+                Row("a", 3, "z", 9),
+                Row("a", 4, "", 2),
+                Row("a", 4, "", 2),
+                Row("b", 1, "h", 5),
+                Row("b", 1, "n", 6),
+                Row("c", 1, "a", 3),
+                Row("c", 1, "z", 9),
+                Row("c", 2, nullStr, 1)
+              )
+            )
+
+            // RangeFrame
+            val existWindowGroupLimitRangeFrame =
+              df.withColumn("sum_value", sum("value").over(window))
+              .limit(1)
+              .queryExecution.optimizedPlan.exists {
+                case _: WindowGroupLimit => true
+                case _ => false
+              }
+            if (threshold == -1) {
+              assert(!existWindowGroupLimitRangeFrame)
+            } else {
+              assert(existWindowGroupLimitRangeFrame)
+            }
+            checkAnswer(df.withColumn("sum_value", sum("value").over(window)).limit(1),
+              Seq(
+                Row("a", 4, "", 8)
+              )
+            )
+            checkAnswer(df.withColumn("sum_value", sum("value").over(window2)).limit(7),
+              Seq(
+                Row("a", 0, "c", 6),
+                Row("a", 1, "x", 6),
+                Row("a", 2, "y", 5),
+                Row("a", 3, "z", 3),
+                Row("a", 4, "", 14),
+                Row("a", 4, "", 14),
+                Row("b", 1, "n", 1)
+              )
+            )
+            checkAnswer(df.withColumn("sum_value", sum("value").over(window3)).limit(11),
+              Seq(
+                Row("a", 0, "c", 11),
+                Row("a", 1, "x", 14),
+                Row("a", 2, "y", 16),
+                Row("a", 3, "z", 20),
+                Row("a", 4, "", 10),
+                Row("a", 4, "", 10),
+                Row("b", 1, "h", 12),
+                Row("b", 1, "n", 13),
+                Row("c", 1, "a", 11),
+                Row("c", 1, "z", 20),
+                Row("c", 2, nullStr, 2)
+              )
+            )
+
+            // Both RowFrame and RangeFrame exist
+            checkAnswer(
+              df.withColumn("sum_value1", sum("value")
+                .over(window.rowsBetween(Window.unboundedPreceding, Window.currentRow)))
+              .withColumn("sum_value2", sum("value")
+                .over(window.rangeBetween(Window.unboundedPreceding, Window.currentRow))).limit(1),
+              Seq(
+                Row("a", 4, "", 4, 8)
```

Comment:
Looks good.

Reply:
Shall we add another group with limit 7 ?

Reply:
Added.